epoxy-processor/src/main/java/com/airbnb/epoxy/KotlinExtensionWriter.kt

```diff
@@ -0,0 +1,158 @@
+package com.airbnb.epoxy
+
+import com.squareup.kotlinpoet.*
+import java.io.*
+import javax.annotation.processing.*
+import javax.lang.model.element.*
+
+private const val KOTLIN_EXTENSION_FILE_NAME = "EpoxyModelKotlinExtensions"
+
+internal class KotlinExtensionWriter(
+        private val processingEnv: ProcessingEnvironment,
+        private val errorLogger: ErrorLogger
+) {
+
+    fun generateExtensionsForModels(generatedModels: List<GeneratedModelInfo>) {
+        val kaptGeneratedDirPath = processingEnv.options[EpoxyProcessor.KAPT_KOTLIN_GENERATED_OPTION_NAME]
+                ?.replace("kaptKotlin", "kapt")
+                ?: run {
+            // Need to change the path because of https://youtrack.jetbrains.com/issue/KT-19097
+
+            // If the option does not exist this is not being processed by kapt,
+            // so we don't need to generate kotlin extensions
+            return
+        }
+
+        generatedModels.groupBy { it.generatedClassName.packageName() }
+                .map { (packageName, models) ->
+                    buildExtensionFile(
+                            packageName,
+                            models)
+                }
+                .forEach {
+                    it.writeTo(File(kaptGeneratedDirPath))
+                }
+    }
+
+    private fun buildExtensionFile(
+            packageName: String,
+            models: List<GeneratedModelInfo>
+    ): KotlinFile {
+        val fileBuilder = KotlinFile.builder(
+                packageName,
+                KOTLIN_EXTENSION_FILE_NAME)
+
+        models.map { buildExtensionForModel(it) }
+                .forEach { fileBuilder.addFun(it) }
+
+        return fileBuilder.build()
+    }
+
+    private fun buildExtensionForModel(model: GeneratedModelInfo): FunSpec {
+        val constructorIsNotPublic = model.constructors
+                .filter { it.params.isEmpty() }
+                .any { !it.modifiers.contains(Modifier.PUBLIC) }
```

Comment:
could be changed to more idiomatic Kotlin:
`Modifier.PUBLIC !in modifiers`

Reply:
great point